pallets/gear/src/tests.rs

```diff
@@ -8008,6 +8067,121 @@ fn system_reservation_gas_allowance_rollbacks() {
     });
 }
 
+#[test]
+fn system_reservation_wait_and_exit_across_executions() {
+    use demo_signal_entry::{HandleAction, WASM_BINARY};
+
+    init_logger();
+    new_test_ext().execute_with(|| {
+        assert_ok!(Gear::upload_program(
+            RuntimeOrigin::signed(USER_1),
+            WASM_BINARY.to_vec(),
+            DEFAULT_SALT.to_vec(),
+            USER_1.encode(),
+            10_000_000_000,
+            0,
+        ));
+
+        let pid = get_last_program_id();
+
+        run_to_block(2, None);
+
+        assert_ok!(Gear::send_message(
+            RuntimeOrigin::signed(USER_1),
+            pid,
+            HandleAction::Wait.encode(),
+            10_000_000_000,
+            0,
+        ));
+
+        let mid_wait = get_last_message_id();
+
+        run_to_block(3, None);
+
+        assert_ok!(GasHandlerOf::<Test>::get_system_reserve(mid_wait));
+
+        assert_ok!(Gear::send_message(
+            RuntimeOrigin::signed(USER_1),
+            pid,
+            HandleAction::Exit.encode(),
+            10_000_000_000,
+            0,
+        ));
+
+        let mid_exit = get_last_message_id();
+
+        run_to_block(4, None);
+
+        assert!(Gear::is_exited(pid));
+        assert!(GasHandlerOf::<Test>::get_system_reserve(mid_wait).is_err());
+        assert!(GasHandlerOf::<Test>::get_system_reserve(mid_exit).is_err());
+
+        MailboxOf::<Test>::clear();
+
+        let mut expiration = None;
+
+        System::events().iter().for_each(|e| {
+            if let MockRuntimeEvent::Gear(Event::MessageWaited {
+                expiration: exp, ..
+            }) = e.event
+            {
+                expiration = Some(exp);
+            }
+        });
+
+        let expiration = expiration.unwrap();
+
+        System::set_block_number(expiration - 1);
+        Gear::set_block_number((expiration - 1).try_into().unwrap());
+
+        run_to_next_block(None);
+
+        // nothing happened after
+        assert!(MailboxOf::<Test>::is_empty(&USER_1));
+    });
+}
+
+#[test]
+fn signal_on_terminated_program() {
```

Comment:
scheme of the test should be:
- msg A is going to init program
- in init msg A sends msg B and waits for reply
- user replies on message B with msg C
- msg C goes into handle_reply section
- in handle reply msg C should: wake msg A, make reservation and wait
- once msg A processed after wake it panics, triggering termination